File: nvpl_blas/f77/dgemm_batch_strided.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DGEMM_BATCH_STRIDED 
!
!******************************************************************************
      program   DGEMM_BATCH_STRIDED_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1      transa, transb
      parameter        (transa='N', transb='N')
      integer          m, n, k, i, j, batch_size
      integer          lda, ldb, ldc
      parameter        (m=2, n=2, k=2, batch_size=2,
     &                  lda=2, ldb=2, ldc=2)
      integer          stridea, strideb, stridec
      parameter        (stridea=lda*k, strideb=ldb*n, stridec=ldc*n)
      double precision alpha, beta
      parameter        (alpha=1.0, beta=1.0)
      double precision a(lda,k,batch_size), b(ldb,n,batch_size),
     &                 c(ldc,n,batch_size)

*       External Subroutines
      external         print_dmatrix, fill_dmatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, m, n, k, lda, ldb, ldc, transa, transb
      print 103, stridea, strideb, stridec, batch_size
      print 101, alpha, beta
      do i = 1, batch_size
          call fill_dmatrix(a(:,:,i),m,k,lda,'F','N')
          call fill_dmatrix(b(:,:,i),k,n,ldb,'F','N')
          call fill_dmatrix(c(:,:,i),m,n,ldc,'F','N')
          call print_dmatrix(a(:,:,i),m,k,lda,'A')
          call print_dmatrix(b(:,:,i),k,n,ldb,'B')
          call print_dmatrix(c(:,:,i),m,n,ldc,'C')
      end do
*
*      Call DGEMM_BATCH_STRIDED subroutine
      call DGEMM_BATCH_STRIDED(transa,transb,m,n,k,alpha,a,lda,
     &     stridea,b,ldb,strideb,beta,c,ldc,stridec,batch_size)
*
      print*
      print 102, 'After DGEMM_BATCH_STRIDED operation:'
      do i = 1, batch_size
          call print_dmatrix(c(:,:,i),m,n,ldc,'C')
      end do

      stop
 99   format('Example: DGEMM_BATCH_STRIDED for the matrix-matrix'
     &       ' multiplication of a batch of matrices')
 100  format('#### args: m=',i1,', n=',i1,', k=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', transA=',a1,', transB=',a1)
 101  format(11x,'alpha=',f4.2,'  beta=',f4.2)
 102  format(A)
 103  format(11x,'stridea=',i1,', strideb=',i1,', stridec=',i1,
     &       ', batch_size=',i1)
      end

File: nvpl_blas/c/dgemm_batch_strided.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dgemm_batch_strided
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 2;
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    double alpha = 1.0f;
    double beta = 1.0f;
    nvpl_int_t batch_size = 2;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    enum CBLAS_TRANSPOSE transB = CblasNoTrans;
    double * A = NULL;
    double * B = NULL;
    double * C = NULL;
    int64_t matrixSizeA = 0;
    int64_t matrixSizeB = 0;
    int64_t matrixSizeC = 0;

    nvpl_int_t rowsA = (transA == CblasNoTrans) ? M : K;
    nvpl_int_t colsA = (transA == CblasNoTrans) ? K : M;
    nvpl_int_t rowsB = (transB == CblasNoTrans) ? K : N;
    nvpl_int_t colsB = (transB == CblasNoTrans) ? N : K;
    nvpl_int_t rowsC = M;
    nvpl_int_t colsC = N;

    matrixSizeA = (int64_t)lda * colsA;
    matrixSizeB = (int64_t)ldb * colsB;
    matrixSizeC = (int64_t)ldc * colsC;
    nvpl_int_t strideA = matrixSizeA;
    nvpl_int_t strideB = matrixSizeB;
    nvpl_int_t strideC = matrixSizeC;
    printf("\nExample: cblas_dgemm_batch_strided for the matrix-matrix multiplication of a batch of matrices\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "transA=%c, transB=%c, order=%c\n", (int64_t)M, (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, transpose_to_char(transA), transpose_to_char(transB), order_to_char(order));
    printf("           strideA=%" PRId64 ", strideB=%" PRId64 ", strideC=%" PRId64 ", batch_size=%" PRId64 "\n",
            (int64_t)strideA, (int64_t)strideB, (int64_t)strideC, (int64_t)batch_size);
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    // allocate memory
    A = (double *)malloc((strideA * (int64_t)batch_size) * sizeof(double));
    B = (double *)malloc((strideB * (int64_t)batch_size) * sizeof(double));
    C = (double *)malloc((strideC * (int64_t)batch_size) * sizeof(double));

    // fill data
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        fill_dmatrix(A + strideA * i, rowsA, colsA, lda, order, Full, CblasNonUnit);
        fill_dmatrix(B + strideB * i, rowsB, colsB, ldb, order, Full, CblasNonUnit);
        fill_dmatrix(C + strideC * i, rowsC, colsC, ldc, order, Full, CblasNonUnit);
    }

    // print input data
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        print_dmatrix(A + strideA * i, rowsA, colsA, lda, "A", order);
        print_dmatrix(B + strideB * i, rowsB, colsB, ldb, "B", order);
        print_dmatrix(C + strideC * i, rowsC, colsC, ldc, "C", order);
    }

    // call cblas_dgemm_batch_strided
    cblas_dgemm_batch_strided(order, transA, transB, M, N, K, alpha, A, lda, strideA,
            B, ldb, strideB, beta, C, ldc, strideC, batch_size);

    // print result
    printf("\nAfter gemm operation:\n");
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        print_dmatrix(C + strideC * i, rowsC, colsC, ldc, "C", order);
    }


    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}
